### ISSUE TITLE
Code does not get generated if project has only Java source files

### ISSUE THREAD

**Issue by rossbacher** [google/ksp#493]
If a project employing ksp has only Java source files the code generated by the processor might not get generate reliably.

Example: 

The Sample app of DeepLinkDispatch (https://github.com/airbnb/DeepLinkDispatch/tree/supportKsp) has only Java sources.
If you run `./gradlew :sample:assemble` it fails to compile, as the ksp Processor does not generate any output files and those are referenced in the project.

Now if you add just one Kotlin class file to the sample project (just create any new class it does not need to be used at all) and rerun the build the files get generated by the processor.

Oddly enough if you go into the source of `DeepLinkProcessor.kt` and change any of the output files filer mode (e.g. from aggregating to isolating) and run the build of the sample app again it works and the output files get generated (even without any Kotlin source file). However if you delete the `build` folder manually and build again the build will fail (until you either add any Kotlin file or change the Filer mode again.

**Comment by ZacSweers** [comment]
I don't think this is really avoidable. KSP hooks into kotlinc's `AnalysisHandlerExtension` API, but kotlinc won't run if there are not Kotlin files. You will need to add a single Kotlin file to make kotlinc run.

We do something similar to avoid java-only projects in general because they break incremental compilation on downstream projects.

```kotlin
/**
 * This type exists solely to force kotlinc to produce incremental compilation information. If we ever add meaningful
 * Kotlin sources to this project, we can then remove this class.
 *
 * TODO this can be removed in AGP 7 when buildConfig can be in kotlin
 */
@Suppress("UnusedPrivateClass")
private abstract class BuildConfigCompilationMarker
```

**Comment by rossbacher** [comment]
That is a good point, I did not look that deeply into it but that makes sense, Should be documented up front though, this drove us crazy hunting this down (especially as it worked when the processor needed to be recompiled (which still confuses me as how would that force a kotlinc run on a downstream project that does not have any kotlin code?)

This is/was mainly a problem in the demo project of DeepLinkDispatch as that was all java before and I did not update/convert the all Java sample app for ksp support.

**Comment by ting-yuan** [comment]
It might be solvable by the `-Xallow-no-source-files` compiler argument. Let me try it and update later.

**Comment by ting-yuan** [comment]
Fixed in #535.

**Comment by rossbacher** [comment]
🙏
